include/nil/crypto3/zk/commitments/polynomial/kzg_v2.hpp

```diff
@@ -0,0 +1,348 @@
+                                set.insert(this->_points[k][i].begin(), this->_points[k][i].end());
+                            }
+                        }
+                        _merged_points = std::vector<typename  KZGScheme::scalar_value_type>(set.begin(), set.end());
+                    }
+
+                    typename math::polynomial<typename KZGScheme::scalar_value_type>
+                    set_difference_polynom(
+                        std::vector<typename KZGScheme::scalar_value_type> merged_points,
+                        std::vector<typename KZGScheme::scalar_value_type> points
+                    ) {
+                        std::sort(merged_points.begin(), merged_points.end());
+                        std::sort(points.begin(), points.end());
+                        std::vector<typename KZGScheme::scalar_value_type> result;
+                        std::set_difference(merged_points.begin(), merged_points.end(), points.begin(), points.end(), std::back_inserter(result));
+                        if (result.size() == 0) {
+                            return typename math::polynomial<typename KZGScheme::scalar_value_type>({{1}});
+                        }
+                        BOOST_ASSERT(this->get_V(result) * this->get_V(points) == this->get_V(merged_points));
+                        return this->get_V(result);
+                    }
+
+                    void update_transcript(std::size_t batch_ind, typename KZGScheme::transcript_type &transcript) {
+                        /* The procedure of updating the transcript is subject to review and change
+                         * #295 */
+
+                        // Push commitments to transcript
+                        transcript(_commitments[batch_ind]);
+
+                        // Push evaluation points to transcript
+                        for( std::size_t i = 0; i < this->_z.get_batch_size(batch_ind); i++){
+                            for( std::size_t j = 0; j < this->_z.get_poly_points_number(batch_ind, i); j++  ) {
+                                nil::marshalling::status_type status;
+                                std::vector<uint8_t> byteblob =
+                                    nil::marshalling::pack<endianness>(this->_z.get(batch_ind, i, j), status);
+                                BOOST_ASSERT(status == nil::marshalling::status_type::success);
+                                transcript(byteblob);
+                            }
+                        }
+
+                        // Push U polynomials to transcript
+                        for (std::size_t i = 0; i < this->_points[batch_ind].size(); i++) {
+                            auto poly = this->get_U(batch_ind, i);
+                            for (std::size_t j = 0; j < poly.size(); ++j) {
+                                nil::marshalling::status_type status;
+                                std::vector<uint8_t> byteblob =
```

Comment:
Same here, we must push the field element into transcript, and if it is not poseidon, then the hash itself will marshall it when needed, not the external code.

Reply:
Made it in https://github.com/NilFoundation/crypto3-zk/pull/313